plutus-metatheory/src/VerifiedCompilation/UCaseOfCase.lagda.md

```diff
@@ -0,0 +1,100 @@
+```
+module VerifiedCompilation.UCaseOfCase where
+
+```
+
+## Imports
+
+```
+open import Data.Nat using (ℕ;zero;suc)
+open import Data.Empty using (⊥)
+open import Scoped using (ScopedTm; Weirdℕ)
+open import Untyped using (_⊢; case; builtin; _·_; force; `; ƛ; delay; con; constr; error)
+open import Untyped.CEK using (BApp; fullyAppliedBuiltin; BUILTIN; stepper; State; Stack)
+open import Evaluator.Base using (maxsteps)
+open import Builtin using (Builtin; ifThenElse)
+open import Data.List using (List; zip; [_])
+open import Utils as U using (Maybe;nothing;just;Either)
+open import RawU using (TmCon)
+open import Data.List.Relation.Unary.All using (All; []; _∷_)
+open import Data.Product using (_×_; proj₁; proj₂)
+open import Relation.Nullary using (Dec; yes; no)
+```
+
+## Translation Relation
+
+This compiler stage only applies to the very specific case where an `IfThenElse` builtin exists in a `case` expression.
+It moves the `IfThenElse` outside and creates two `case` expressions with each of the possible lists of cases. 
+
+This translation relation has a constructor for the specific case, and then inductive constructors for everything else
+to traverse the ASTs.
+
+```
+data _⊢̂_⊳̂_ (X : Set) : (X ⊢) → (X ⊢) → Set where
+--   refl : ∀ {X : Set} {x : X ⊢} → X ⊢̂ x ⊳̂ x
+   caseofcase : ∀ {b : X ⊢} {tn tn' fn fn' : ℕ} {alts alts' tt ft tt' ft' : List (X ⊢)} 
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip alts alts') -- recursive translation for the other case patterns 
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip tt tt') -- recursive translation for the true branch
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip ft ft') -- recursive translation for the false branch
+                ----------------------
+                → X ⊢̂
+                       (case ((((force (builtin ifThenElse)) · b) · (constr tn tt)) · (constr fn ft)) alts)
```

Comment:
No translation for `b`? It can be an arbitrary expression.

Why does it have to be `constr` rather than an arbitrary expression?

Also `ifThenElse` is strict, since the language is strict and all builtins are strict, hence what we normally do is represent the branches of `ifThenElse` as functions or `delay`ed values and that is what ends up being in the actual validators, so as stated this translation relation isn't very useful.

Reply:
This is where I was wanting your input: the comments on the Haskell, and my understanding of the Haskell suggest that the expressions have to be `constr`. We (Ana and Jacco really) came up with a reasonable explanation as to why this was important, but perhaps you can explain it better (and we should update the comments)?

Reply:
This is also what the tests seem to document: https://github.com/IntersectMBO/plutus/blob/f1eae1c5c8def282d91cf38a58c1493e09e1fa70/plutus-core/untyped-plutus-core/test/Transform/Simplify.hs#L48

Reply:
Weird! I'll look into it.

Reply:
> the comments on the Haskell, and my understanding of the Haskell suggest that the expressions have to be `constr`

Yes, it turns out you're right. Was discussed [here](https://github.com/IntersectMBO/plutus/pull/5337#discussion_r1199215315).

> Also `ifThenElse` is strict, since the language is strict and all builtins are strict, hence what we normally do is represent the branches of `ifThenElse` as functions or delayed values and that is what ends up being in the actual validators, so as stated this translation relation isn't very useful.

This does look like a bug with the Haskell code, I've talked with Yura about it and he's going to take a look.